### PR TITLE
policy: Add framework for async event completion notifications.

### PR DESCRIPTION
**Pull request by jrajahalme**
Add a framework for asynchronous proxy update completion notifications and use it for Envoy.

### CODE REVIEW

**Comment by jrajahalme** [comment]
The envoy pkg unit test shows this in action:
```
level=debug msg="Envoy: Bootstrap: node:<id:\"envoy1\" cluster:\"cluster1\" build_version:\"version1\" > static_resources:<clusters:<name:\"cluster1\" type:ORIGINAL_DST connect_timeout:<seconds:1 > lb_policy:ORIGINAL_DST_LB cleanup_interval:<seconds:1 nanos:500000000 > auto_http2:true > clusters:<name:\"ldsCluster\" connect_timeout:<seconds:1 > hosts:<pipe:<path:\"/tmp/check-8674665223082153551/0/lds.sock\" > > http2_protocol_options:<> > clusters:<name:\"rdsCluster\" connect_timeout:<seconds:1 > hosts:<pipe:<path:\"/tmp/check-8674665223082153551/0/rds.sock\" > > http2_protocol_options:<> > > dynamic_resources:<lds_config:<api_config_source:<api_type:GRPC cluster_name:\"ldsCluster\" > > > admin:<access_log_path:\"/dev/null\" address:<socket_address:<address:\"127.0.0.1\" port_value:9942 > > > "
level=debug msg="Envoy: Starting {<nil> /tmp/check-8674665223082153551/0/cilium-envoy.log /tmp/check-8674665223082153551/0/access_log.sock 0xc42049bcb0 /tmp/check-8674665223082153551/0/lds.sock <nil> /tmp/check-8674665223082153551/0/rds.sock <nil> 0xc4203132e0}"
level=debug msg="Envoy: Started."
level=info msg="Envoy: Process started at pid 23456"
level=debug msg="Envoy: addListener listener1"
level=debug msg="Envoy: AddCompletion: addListener listener1"
level=debug msg="Envoy: addListener listener2"
level=debug msg="Envoy: AddCompletion: addListener listener2"
level=debug msg="Envoy: addListener listener3"
level=debug msg="Envoy: AddCompletion: addListener listener3"
level=debug msg="Waiting for proxy updates to complete..."
level=debug msg="Envoy: LDS Stream DiscoveryRequest: node:<id:\"envoy1\" cluster:\"cluster1\" build_version:\"356598fb8ebd2217a92311266a0f1a833bc52c3e/Modified/DEBUG\" > type_url:\"type.googleapis.com/envoy.api.v2.Listener\" "
level=debug msg="Envoy: LDS detected Envoy (re)start "
level=debug msg="Envoy: Starting stream handler for: LDS"
level=info msg="Envoy: Access log connection opened"
level=debug msg="Envoy: LDS Stream DiscoveryRequest: version_info:\"4\" node:<id:\"envoy1\" cluster:\"cluster1\" build_version:\"356598fb8ebd2217a92311266a0f1a833bc52c3e/Modified/DEBUG\" > type_url:\"type.googleapis.com/envoy.api.v2.Listener\" "
level=debug msg="Envoy: LDS ACK received: 4"
level=debug msg="Envoy: RDS Stream DiscoveryRequest node:<id:\"envoy1\" cluster:\"cluster1\" build_version:\"356598fb8ebd2217a92311266a0f1a833bc52c3e/Modified/DEBUG\" > resource_names:\"RDS_listener3_3\" type_url:\"type.googleapis.com/envoy.api.v2.RouteConfiguration\" "
level=debug msg="Envoy: RDS_listener3_3 detected Envoy (re)start "
level=debug msg="Envoy: Starting stream handler for: RDS_listener3_3"
level=debug msg="Envoy: RDS Stream DiscoveryRequest node:<id:\"envoy1\" cluster:\"cluster1\" build_version:\"356598fb8ebd2217a92311266a0f1a833bc52c3e/Modified/DEBUG\" > resource_names:\"RDS_listener2_2\" type_url:\"type.googleapis.com/envoy.api.v2.RouteConfiguration\" "
level=debug msg="Envoy: RDS_listener2_2 detected Envoy (re)start "
level=debug msg="Envoy: Starting stream handler for: RDS_listener2_2"
level=debug msg="Envoy: RDS Stream DiscoveryRequest node:<id:\"envoy1\" cluster:\"cluster1\" build_version:\"356598fb8ebd2217a92311266a0f1a833bc52c3e/Modified/DEBUG\" > resource_names:\"RDS_listener1_1\" type_url:\"type.googleapis.com/envoy.api.v2.RouteConfiguration\" "
level=debug msg="Envoy: RDS_listener1_1 detected Envoy (re)start "
level=debug msg="Envoy: Starting stream handler for: RDS_listener1_1"
level=debug msg="Envoy: RDS Stream DiscoveryRequest version_info:\"1\" node:<id:\"envoy1\" cluster:\"cluster1\" build_version:\"356598fb8ebd2217a92311266a0f1a833bc52c3e/Modified/DEBUG\" > resource_names:\"RDS_listener3_3\" type_url:\"type.googleapis.com/envoy.api.v2.RouteConfiguration\" "
level=debug msg="Envoy: RDS_listener3_3 ACK received: 1"
level=debug msg="Envoy: RDS_listener3_3 addListener listener3 ACK, time left: 9.81076055s"
level=debug msg="completions.Completed: true"
level=debug msg="Envoy: RDS Stream DiscoveryRequest version_info:\"1\" node:<id:\"envoy1\" cluster:\"cluster1\" build_version:\"356598fb8ebd2217a92311266a0f1a833bc52c3e/Modified/DEBUG\" > resource_names:\"RDS_listener1_1\" type_url:\"type.googleapis.com/envoy.api.v2.RouteConfiguration\" "
level=debug msg="Envoy: RDS_listener1_1 ACK received: 1"
level=debug msg="Envoy: RDS_listener1_1 addListener listener1 ACK, time left: 9.809043469s"
level=debug msg="completions.Completed: true"
level=debug msg="Envoy: RDS Stream DiscoveryRequest version_info:\"1\" node:<id:\"envoy1\" cluster:\"cluster1\" build_version:\"356598fb8ebd2217a92311266a0f1a833bc52c3e/Modified/DEBUG\" > resource_names:\"RDS_listener2_2\" type_url:\"type.googleapis.com/envoy.api.v2.RouteConfiguration\" "
level=debug msg="Envoy: RDS_listener2_2 ACK received: 1"
level=debug msg="Envoy: RDS_listener2_2 addListener listener2 ACK, time left: 9.809517644s"
level=debug msg="completions.Completed: true"
level=debug msg="Proxy updates completed in 190.127394ms"
level=debug msg="Envoy: updateListener listener2"
level=debug msg="Envoy: AddCompletion: updateListener listener2"
level=debug msg="Waiting for proxy updates to complete..."
level=debug msg="Envoy: RDS Stream DiscoveryRequest version_info:\"2\" node:<id:\"envoy1\" cluster:\"cluster1\" build_version:\"356598fb8ebd2217a92311266a0f1a833bc52c3e/Modified/DEBUG\" > resource_names:\"RDS_listener2_2\" type_url:\"type.googleapis.com/envoy.api.v2.RouteConfiguration\" "
level=debug msg="Envoy: RDS_listener2_2 ACK received: 2"
level=debug msg="Envoy: RDS_listener2_2 updateListener listener2 ACK, time left: 9.999186424s"
level=debug msg="completions.Completed: true"
level=debug msg="Proxy updates completed in 834.113µs"
level=debug msg="Envoy: updateListener listener1"
level=debug msg="Envoy: AddCompletion: updateListener listener1"
level=debug msg="Waiting for proxy updates to complete..."
level=debug msg="Envoy: RDS Stream DiscoveryRequest version_info:\"2\" node:<id:\"envoy1\" cluster:\"cluster1\" build_version:\"356598fb8ebd2217a92311266a0f1a833bc52c3e/Modified/DEBUG\" > resource_names:\"RDS_listener1_1\" type_url:\"type.googleapis.com/envoy.api.v2.RouteConfiguration\" "
level=debug msg="Envoy: RDS_listener1_1 ACK received: 2"
level=debug msg="Envoy: RDS_listener1_1 updateListener listener1 ACK, time left: 9.99912906s"
level=debug msg="completions.Completed: true"
level=debug msg="Proxy updates completed in 906.599µs"
level=debug msg="Envoy: removeListener listener3"
level=debug msg="Envoy: AddCompletion: removeListener listener3"
level=debug msg="Waiting for proxy updates to complete..."
level=debug msg="Envoy: Stream handler stopped: RDS_listener3_3"
level=debug msg="Envoy: LDS Stream DiscoveryRequest: version_info:\"5\" node:<id:\"envoy1\" cluster:\"cluster1\" build_version:\"356598fb8ebd2217a92311266a0f1a833bc52c3e/Modified/DEBUG\" > type_url:\"type.googleapis.com/envoy.api.v2.Listener\" "
level=debug msg="Envoy: LDS ACK received: 5"
level=debug msg="Envoy: LDS removeListener listener3 ACK, time left: 9.998239401s"
level=debug msg="completions.Completed: true"
level=debug msg="Proxy updates completed in 1.793099ms"
level=debug msg="Envoy: addListener listener3"
level=debug msg="Envoy: AddCompletion: addListener listener3"
level=debug msg="Waiting for proxy updates to complete..."
level=debug msg="Envoy: LDS Stream DiscoveryRequest: version_info:\"6\" node:<id:\"envoy1\" cluster:\"cluster1\" build_version:\"356598fb8ebd2217a92311266a0f1a833bc52c3e/Modified/DEBUG\" > type_url:\"type.googleapis.com/envoy.api.v2.Listener\" "
level=debug msg="Envoy: LDS ACK received: 6"
level=debug msg="Envoy: RDS Stream DiscoveryRequest node:<id:\"envoy1\" cluster:\"cluster1\" build_version:\"356598fb8ebd2217a92311266a0f1a833bc52c3e/Modified/DEBUG\" > resource_names:\"RDS_listener3_5\" type_url:\"type.googleapis.com/envoy.api.v2.RouteConfiguration\" "
level=debug msg="Envoy: RDS_listener3_5 detected Envoy (re)start "
level=debug msg="Envoy: Starting stream handler for: RDS_listener3_5"
level=debug msg="Envoy: RDS Stream DiscoveryRequest version_info:\"1\" node:<id:\"envoy1\" cluster:\"cluster1\" build_version:\"356598fb8ebd2217a92311266a0f1a833bc52c3e/Modified/DEBUG\" > resource_names:\"RDS_listener3_5\" type_url:\"type.googleapis.com/envoy.api.v2.RouteConfiguration\" "
level=debug msg="Envoy: RDS_listener3_5 ACK received: 1"
level=debug msg="Envoy: RDS_listener3_5 addListener listener3 ACK, time left: 9.970685484s"
level=debug msg="completions.Completed: true"
level=debug msg="Proxy updates completed in 29.470214ms"
level=debug msg="Envoy: removeListener listener3"
level=debug msg="Envoy: AddCompletion: removeListener listener3"
level=debug msg="Envoy: Stream handler stopped: RDS_listener3_5"
level=info msg="Envoy: Stopping process 23456"
level=debug msg="Envoy: Stream handler stopped: RDS_listener2_2"
2018/01/31 19:27:38 transport: http2Server.HandleStreams failed to read frame: read unix /tmp/check-8674665223082153551/0/rds.sock->@: use of closed network connection
level=debug msg="Envoy: Stream handler stopped: RDS_listener1_1"
2018/01/31 19:27:38 transport: http2Server.HandleStreams failed to read frame: read unix /tmp/check-8674665223082153551/0/lds.sock->@: use of closed network connection
level=warning msg="Envoy: LDS Send() failed" error="rpc error: code = 13 desc = transport is closing"
level=debug msg="Envoy: LDS removeListener listener3 NACK, time left: 9.997355602s"
level=debug msg="completions.Completed: false"
level=debug msg="Envoy: Stream handler stopped: LDS"
level=debug msg="Envoy admin response to quitquitquit: OK\n"
level=debug msg="Waiting for proxy updates to complete..."
level=debug msg="Proxy updates failed: 1 proxy state changes failed"
level=info msg="Envoy: Access log closing"
level=info msg="Envoy: Sleeping for 100ms before respawning." error="<nil>"
PASS: envoy_test.go:75: EnvoySuite.TestEnvoy	0.252s
OK: 1 passed
--- PASS: Test (0.25s)
PASS
coverage: 66.5% of statements
ok  	github.com/cilium/cilium/pkg/envoy	0.307s
```

The last failure above is intentional.

**Comment by jrajahalme** [comment]
test-me-please

File: pkg/policy/policy.go
Comment (rlenglet) [inline pkg/policy/policy.go:36]: pooling -> polling (?)

File: pkg/envoy/streamctrl.go
Comment (rlenglet) [inline pkg/envoy/streamctrl.go]:
```
for _, completion := range ctrl.completions {
    ...
}
```

File: pkg/policy/policy.go
Comment (rlenglet) [inline pkg/policy/policy.go:39]: Instead of explicitly using a timeout, why not pass a `context.Context` to every calls that currently takes a `CompletionContainer`? Using a context will allow your streamctrl logic to support both explicit cancelations from callers and timeouts.
Instead of directly dealing with timeouts, you'd only need to check whether the context is still valid with `ctx.Err() == nil` to determine whether to continue waiting for an ACK.

Comment (rlenglet) [reply pkg/policy/policy.go:39]: Also, all functions that currently take a `CompletionContainer` as a parameter only either pass it down the call stack, or call `AddCompletion` only once. I'd directly pass a `Completion` parameter instead of a `CompletionContainer`.

It may also be worth trying to implement the Future pattern, because that's really what you try to do here. A future may simply be modeled with a `chan bool` here. The callee would write at most one value into the channel and close it, once the version is ACKed/NACKed by the proxy.
The functions' interfaces would then look like:
```
func (s *LDSServer) updateListener(ctx context.Context, ..., out chan<- bool) {
    ...
}
```

If you wanted to define a fancy type to wrap the channel details, you could pass a `Completion` to do the job, and implement it so:
```
type Completion func(success bool)

func NewCompletion(ctx Context, out chan<- bool) Completion {
    return func(success bool) {
        select {
        case <-ctx.Done():
            // Ignore.
        case out <- success:
        }
        close(out)
    }
}

func (s *LDSServer) updateListener(ctx context.Context, ..., completion Completion) {
    ...
}
```

File: pkg/policy/policy.go
Comment (rlenglet) [inline pkg/policy/policy.go:32]: This could simply be defined as a func:
```
type Completion func(success bool)
```

**Review by rlenglet — CHANGES_REQUESTED** [review]
I have not reviewed all yet. I have general comments about the design of Completion / CompletionContainer.

File: pkg/policy/policy.go
Comment (rlenglet) [reply pkg/policy/policy.go:39]: Now I think you'll always want to combine a completion with a context, so it may be better to define the interface so:
```
type Completion interface {
    Context() context.Context
    Completed(success bool)
}
```

And just pass only that around, e.g.:
```
func (e *Envoy) UpdateListener(name string, l7rules policy.L7DataMap, completion policy.Completion) {
    ...
}
```

File: pkg/envoy/streamctrl.go
Comment (jrajahalme) [reply pkg/envoy/streamctrl.go]: Does that support removal of entries in the slice while iterating?

File: pkg/policy/policy.go
Comment (jrajahalme) [reply pkg/policy/policy.go:32]: An interface carries an object reference with it, while a function pointer does not. Multiple different endpoints may request completion callbacks in parallel from the same stream (in case of LDS).

File: pkg/policy/policy.go
Comment (jrajahalme) [reply pkg/policy/policy.go:36]: Pooling is the right spelling here :-)

Multiple completions can be added to the same container, so that the caller can efficiently wait for multiple compoletions at one time.

File: pkg/policy/policy.go
Comment (jrajahalme) [reply pkg/policy/policy.go:39]: Part of my design was to both:
- Allow passing CompletionContainer as nil to support callers who may not be interested in the completion signal
- Allow implementations that are always synchronous to ignore the CompletionContainer

Furthermore, my implementation allows:
- the implementation of a Completion interface package a single WaitGroup
- and share that same instance for multiple Completions in the same CompletionsContainer

This allows for a potentially large number of completions to be waited for with a single WaitGroup.Wait().

To do the same with a single channel without blocking writers I'd have to have a buffered channel for some max number of completions?

Comment (jrajahalme) [reply pkg/policy/policy.go:39]: I'll have to look at the timeout.. Maybe it would be better to have an overall daedline for the whole policy update operartion rather than individual timeouts on each completion. If so the Context() could be a member in the CompletionContainer interface. I guess context.Context must be passed by reference for a cancellation to pass through?

**Review by jrajahalme — COMMENTED** [review]
Thanks for the review! See my responses :-)

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
The extension to proxy API being defined in this PR is fairly simple, consisting of two simple interfaces:

```
type Completion interface {
	Completed(bool)
}

type CompletionContainer interface {
	AddCompletion() (Completion, time.Duration)
}
```

Each proxy API call that can potentially be asynchronous takes the `CompletionContainer` as a new parameter, which can be `nil` if the caller does not care for asynchronous completion notifications.
Synchronous proxy implementations can likewise ignore the new parameter. These allow for minimal complexity for callers/callees that do not need the new functionality.

These interfaces also divide the burden of implemtation between the caller and callee. In this PR the caller needs to track completions of parallel regenerations of proxies for multiple endpoints. Currently this is implemented with a WaitGroup, but the proxy implementations need not know that.

Similarly, a running envoy proxy package consists of a number of parallel gRPC stream handlers, some of which are endpoint specific (RDS, at least with the Envoy version we are using), while others are accessed by multiple endpoint regenerations in parallel (LDS). This requires each handler to be capable of tracking multiple completions at the same time. The caller (i.e., the policy regeneration code) does not need to know that, however.

A nuance in the implementation is that the policy regeneration code initializes a new ProxyCompletions object when it starts. Completions are then collected on this object, which can then be used to wait for completion of all the parallel asynchronous events. In a case of an error, however, policy regeneration code simply abandons the allocated ProxyCompletions object, i.e., it does not need to wait for the completion is something else fails. The gRPC handlers will still signal completions, after which they release their reference to the Completion interface, causing the object to be freed.

**Comment by jrajahalme** [comment]
Current implementation is a bit fragile in the sense that a bug in a proxy package (missing Completed() callback) will leave the calling thread blocked. A more robust implementation would have the timeout implemented at the caller side. However the tradeoff is that we don't want such bugs to exist, so it may be better to make it obvious when/if this is happening.

**Comment by jrajahalme** [comment]
test-me-please

File: pkg/envoy/streamctrl.go
Comment (nebril) [inline pkg/envoy/streamctrl.go]: This is a weird construction IMO, you are effectively shrinking `ctrl.completions` slice until you processed all entries.

If `ctrl.completions` can be changed after going through this loop I would suggest to rewrite this whole loop as follows, to avoid somewhat convoluted slice filtering:
```
	tmpCompletions := make([]versionCompletion, 0, len(ctrl.completions))
	for _, comp := range ctrl.completions {

		result := success

		if version >= comp.version {
			// Ack or Nack for this version OR later received
		} else if now.After(comp.expiration) {
			// Timed out, return failure.
			result = false
		} else {
			// retain the entry
			append(tmpCompletions, comp)
		}

		res := "NACK"
		if result {
			res = "ACK"
		}
		log.Debug("Envoy: ", ctrl.name, " ", comp.msg, " ", res, ", time left: ", comp.expiration.Sub(now))

		comp.completion.Completed(result)
	}
	ctrl.completions = tmpCompletions
```
Hope this makes sense.

**Review by nebril — CHANGES_REQUESTED** [review]
One loop rewrite suggestion inline, please dismiss review if you don't agree with it.

File: pkg/envoy/streamctrl.go
Comment (jrajahalme) [reply pkg/envoy/streamctrl.go]: Makes sense, even though I don't like the fact that range loop copies the object into a temporary (here `comp`). However, in most cases we do this after a successful Envoy transaction, so it makes no difference in practice.

I also did not include the preallocation of the temporary array, as with that the array never shrinks. While we are not dealing with large arrays, IMO it is a bad habit. Again, in most cases I expect we don't need to retain entries, so this does not make a difference either.

I also had to keep the `continue` to not call completion when an entry is retained.

I do like that this is more readable now!

**Comment by jrajahalme** [comment]
test-me-please

**Comment by rlenglet** [comment]
As we discussed offline, it seems that passing a `Context` and a callback function pointer would be the best design:
```
func Dosomething(ctx context.Context, completed func(), ...) {
    ...
}
```

**Comment by rlenglet** [comment]
The caller will create the context with a timeout:
```
ctx, cancel := context.WithTimeout(..., 15 * time.Second)
```

**Comment by rlenglet** [comment]
```
func (e *Endpoint) WaitForProxyCompletions(ctx context.Context, cancel context.CancelFunc) error {
    defer cancel()
    allDone := make(chan strict{})
    go func() {
        defer close(allDone)
        e.ProxyCompletions.wg.Wait()
    }
    select {
    case <-ctx.Done():
    case <-allDone:
    }
    return ctx.Err()
 }
```

**Comment by rlenglet** [comment]
The problem now becomes: how to unblock the goroutine that calls `wg.Wait`? There’s no safe way to unblock a waitgroup wait on the caller side.
I can think of a few solutions:
1) require every caller to call the callback function on completion or when the context times out
2) replace the waitgroup with a channel that is written into by the callback functions; this has the problem it will be hard to know when to close that channel without risking a panic
3) replace the waitgroup with a more open implementation, with a mutex, a counter, and a condition (`Cond`); callback functions would acquire the mutex, decrement the counter, and broadcast the condition; the caller side now becomes:

```
func (e *Endpoint) WaitForProxyCompletions(ctx context.Context, cancel context.CancelFunc) error {
    defer cancel()
    allDone := make(chan strict{})
    go func() {
        defer close(allDone)
        e.ProxyCompletions.locker.Lock()
        defer e.ProxyCompletions.locker.Unlock()
        for ctx.Err() == nil && e.ProxyCompletions.count > 0 {
            e.ProxyCompletions.condition.Wait()
        }
    }()
    select {
    case <-ctx.Done():
    case <-allDone:
    }
    // Just in case count > 0, wake up the goroutine,
    // in order to properly close the allDone channel and exit.
    e.ProxyCompletions.condition.Broadcast()
    return ctx.Err()
 }
```

File: pkg/envoy/envoy_test.go
Comment (tgraf) [inline pkg/envoy/envoy_test.go:41]: Typo in your email in the comment message

File: pkg/endpoint/policy.go
Comment (tgraf) [inline pkg/endpoint/policy.go]: This is very fragile as it depends on the above defer to be called afterwards. Can you do one defer function which sets ProxyCompletions to nil and then unlocks BuildMutex so it's clear that there is a dependency?

**Review by tgraf — CHANGES_REQUESTED** [review]
Everybody please remember that this is an expedited bug which is currently blocking several other bugs. I don't mind long term design discussions but was there anything in the original implementation that did *not* work and requires a re-design or can we merge the original approach and then refine later?

**Review by nebril — APPROVED** [review]

File: pkg/endpoint/policy.go
Comment (jrajahalme) [reply pkg/endpoint/policy.go]: Good point, with one defer the order is more explicit and it becomes a bit more efficient as well.

Comment (jrajahalme) [reply pkg/endpoint/policy.go]: Good point, with one defer the order is more explicit and it becomes a bit more efficient as well.

File: pkg/envoy/envoy_test.go
Comment (jrajahalme) [reply pkg/envoy/envoy_test.go:41]: Fixed.

**Comment by jrajahalme** [comment]
> On Feb 5, 2018, at 2:20 AM, Thomas Graf <notifications@github.com> wrote:
> 
> @tgraf requested changes on this pull request.
> 
> Everybody please remember that this is an expedited bug which is currently blocking several other bugs. I don't mind long term design discussions but was there anything in the original implementation that did not work and requires a re-design or can we merge the original approach and then refine later?
> 
> 

I agree, we should merge this as-is (after the fixes to smaller issues Thomas pointed out), and leave the interface refinement for a later exercise, which can be picked up by anyone who wants to work on it :-)

  Jarno

**Comment by jrajahalme** [comment]
test-me-please

File: daemon/daemon.go
Comment (manalibhutiyani) [inline daemon/daemon.go:130]: Why do we need to pass `e` twice, as well as `e.ProxyID(l4)` separately?

**Review by rlenglet — APPROVED** [review]
OK with addressing the interface design in a subsequent PR.

File: daemon/daemon.go
Comment (jrajahalme) [reply daemon/daemon.go:130]: 'e', endpoint is being passed in different roles (literally different interfaces), the called function does not (need to) know that they are the same 'e':

```
func (p *Proxy) CreateOrUpdateRedirect(l4 *policy.L4Filter, id string, source ProxySource,
	notifier accesslog.LogRecordNotifier, completions policy.CompletionContainer) (Redirect, error)
```

**Review by manalibhutiyani — APPROVED** [review]

**Review by tgraf — APPROVED** [review]

**Pull request merged**
